### PR TITLE
Feature/track list

### PR DESCRIPTION
**Pull request by oszydlowski**

### CODE REVIEW

File: AIM/AIM.xcodeproj/xcuserdata/oskarszydlowski.xcuserdatad/xcschemes/xcschememanagement.plist [hunk 1]
```diff
@@ -7,7 +7,7 @@
 		<key>AIM.xcscheme</key>
 		<dict>
 			<key>orderHint</key>
-			<integer>0</integer>
+			<integer>2</integer>
```
Comment (SebastianOsinski) [inline AIM/AIM.xcodeproj/xcuserdata/oskarszydlowski.xcuserdatad/xcschemes/xcschememanagement.plist:10]: This whole file should be in .gitignore

File: AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    
+    func parse(completion: @escaping OnAirInfoCompletion) {
+        parserCompletion = completion
+        parser.parse()
+    }
+    
+    // MARK: - XMLParserDelegate
+    
+    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String] = [:]) {
+        if elementName == "epgItem" {
```
Comment (SebastianOsinski) [inline AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.swift]: All strings should be extracted to constants, especially `playoutItem` as it is used more than one time

File: AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                    }
+                }
+            }
+        } else if elementName == "playoutItem" {
+            track = Track.fromAttributeDict(attributeDict: attributeDict)
+        }
+    }
+    
+    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
+        if elementName == "playoutItem" {
```
Comment (SebastianOsinski) [inline AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.swift]: `if elementName == playoutItemElementName, let track = track {`

File: AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+//
+//  Created by Oskar Szydlowski on 01.03.2018.
+//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
+//
+
+import Foundation
+
+typealias OnAirInfoCompletion = (OnAirInfo) -> ()
+
+class XMLOnAirInfoGateway: NSObject, XMLParserDelegate {
```
Comment (SebastianOsinski) [inline AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.swift]: This should be hidden behind protocol. Also outside world shouldn't know about any parsing, i.e `parse(completion)` should be renamed

File: AIM/AIM/Presentation/Presenters/TrackListPresenter.swift [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import UIKit
+
+class TrackListPresenter {
+    
+    weak var view: TrackListView?
+    
+    private let connector: TrackListConnector
+    private let useCaseFactory: UseCaseProducing
+    
+    var onAirInfo: OnAirInfo?
```
Comment (SebastianOsinski) [inline AIM/AIM/Presentation/Presenters/TrackListPresenter.swift]: Should be private

File: AIM/AIM/UI/Views/StationInfoView.swift [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            descriptionLabel.topAnchor.constraint(equalTo: nameLabel.bottomAnchor, constant: 10.0),
+            descriptionLabel.leadingAnchor.constraint(equalTo: nameLabel.leadingAnchor),
+            descriptionLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -5.0),
+            descriptionLabel.bottomAnchor.constraint(equalTo: separator.topAnchor)
+        ])
+        
+        descriptionLabel.numberOfLines = 0
+    }
+    
+    func displayImage(_ URL: URL) {
```
Comment (SebastianOsinski) [inline AIM/AIM/UI/Views/StationInfoView.swift:79]: `url` instead of `URL`

File: AIM/AIM/UI/Views/StationInfoView.swift [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        imageView.kf.setImage(with: URL)
```
Comment (SebastianOsinski) [inline AIM/AIM/UI/Views/StationInfoView.swift]: Kingfisher should be hidden behind simple extension on imageView, `imageView.setImage`

File: AIM/AIM/UI/Views/TrackCell.swift [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    
+    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        selectionStyle = .none
+        setupView()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
```
Comment (SebastianOsinski) [inline AIM/AIM/UI/Views/TrackCell.swift:36]: There is no preapreForReuse implemented. It should cancel image fetching task request. Kingfisher returns `RetrieveImageTask` which you should store in cell and use it to cancel fetching. Whole extension for imageview can look like this:

```swift
import UIKit
import Kingfisher

protocol ImageTask {
    func cancel()
}

extension RetrieveImageTask: ImageTask { }

extension UIImageView {

    @discardableResult func setImage(with url: URL?) -> ImageTask {
        return kf.setImage(with: url)
    }
}

```

File: AIM/AIM/UI/Views/TrackListViewController.swift [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    }
+    
+    // MARK: - UITableViewDelegate
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return rowHeight
+    }
+    
+    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        guard let _ = presenter.onAirInfo?.stationInfo else {
```
Comment (SebastianOsinski) [inline AIM/AIM/UI/Views/TrackListViewController.swift]: view shouldn't 'inspect' presenter like that. This breaks law of Demeter. Presenter should provide this with some variable or method, e.g. `isHeaderVisible` or sth similiar

File: AIM/AIM/UI/Views/TrackListViewController.swift [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            return 0.0
+        }
+        
+        return rowHeight
+    }
+    
+    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        guard let _ = presenter.onAirInfo?.stationInfo else {
```
Comment (SebastianOsinski) [inline AIM/AIM/UI/Views/TrackListViewController.swift]: Ditto

File: AIM/AIM/UI/Views/TrackListViewController.swift [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        }
+        
+        let headerView = StationInfoView()
+        presenter.configureHeaderView(headerView)
+        return headerView
+    }
+    
+    //MARK: - UITableViewDataSource
+    
+    func numberOfSections(in tableView: UITableView) -> Int {
```
Comment (SebastianOsinski) [inline AIM/AIM/UI/Views/TrackListViewController.swift]: Not needed

**Review by SebastianOsinski — CHANGES_REQUESTED** [review]

File: AIM/AIM.xcodeproj/xcuserdata/oskarszydlowski.xcuserdatad/xcschemes/xcschememanagement.plist (hunk 1 above)
Comment (oszydlowski) [reply AIM/AIM.xcodeproj/xcuserdata/oskarszydlowski.xcuserdatad/xcschemes/xcschememanagement.plist:10]: I'll update gitignore later, have to finish sth

**Review by SebastianOsinski — APPROVED** [review]

**Review by mateusznadolski — APPROVED** [review]

**Pull request merged**